### PR TITLE
New effect commands with 'rand' support

### PR DESCRIPTION
**Pull request by islemaster**
Part of the fix for https://github.com/code-dot-org/dance-party/issues/469

Introduces `setBackgroundEffectWithPalette` and `setForegroundEffectExtended` which are pass-throughs to the existing functions, but exist so that we can create a new block instead of updating the old one for backwards-compatability.

Adds a random effect key 'rand' that, if passed, will cause a random effect to be selected from a preselected list appropriate to background/foreground as appropriate.

### CODE REVIEW

File: src/p5.dance.js [hunk 1]
```diff
@@ -1030,6 +1036,16 @@ module.exports = class DanceParty {
...
   }
 
+  /**
+   * Randomly pick one element out of an array.
+   * @param {Array.<T>} collection
+   * @returns {T}
+   * @private
+   */
+  sample_(collection) {
+    return collection[Math.floor(this.p5_.random(0, collection.length))];
```
Comment (uponthesun) [inline src/p5.dance.js]: Slightly simpler to use the randomInt function already defined in this file?

File: src/api.js [hunk 2]
```diff
@@ -6,12 +6,18 @@ module.exports = class DanceAPI {
...
+      setBackgroundEffect: (effect, palette) => {
+        nativeAPI.setBackgroundEffect(effect.toString(), palette);
+      },
+      setBackgroundEffectWithPalette: (effect, palette) => {
+        nativeAPI.setBackgroundEffect(effect.toString(), palette);
       },
       setForegroundEffect: effect => {
         nativeAPI.setForegroundEffect(effect.toString());
       },
+      setForegroundEffectExtended: (effect) => {
```
Comment (uponthesun) [inline src/api.js]: Do we need this new function in the API if the signature / behavior is the same? Can't both the old and new blocks call the same existing function ?

File: src/constants.js [hunk 3]
```diff
@@ -29,4 +29,39 @@ module.exports = {
     {name: "XOle", mirror: false, shortBurst: true},
     {name: "XSlide", mirror: false, shortBurst: true},
   ],
+  RANDOM_EFFECT_KEY: 'rand',
+  BACKGROUND_EFFECTS: [
```
Comment (uponthesun) [inline src/constants.js]: Is it possible / realistic that this list gets out of sync and contains a background which doesn't exist yet / no longer exists? Where's the authoritative source of implemented backgrounds? If so, is it possible to fail gracefully if we randomly pick one which doesn't exist?

File: test/unit/apiTest.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
+const sinon = require('sinon');
+const test = require('tape');
+const DanceAPI = require('../../src/api');
+
+const FAKE_EFFECT = 'fake-effect-name';
+const FAKE_PALETTE = 'fake-palette-name';
+
+// DEPRECATED
```
Comment (uponthesun) [inline test/unit/apiTest.js]: Add these comments to the api source file as well?

File: src/api.js (hunk 2 above)
Comment (islemaster) [reply src/api.js]: @ryansloan mentioned there was some reason we can't have two different blocks that point to the same function.  I don't know the details, but providing an alias function in the API seemed like a good workaround.

I'm not actually sure what that limitation is - we use the full block name when serializing to XML (`<block type="Dancelab_setBackgroundEffect">`), so there must be some other place we depend on the `func` being unique across all loaded blocks.

File: src/constants.js (hunk 3 above)
Comment (islemaster) [reply src/constants.js]: That's a great point.  `Effects.js` contains the actual implemented effects.  I'll union this list of 'approved' effects for random selection with the set of implemented effects before we pick one.

File: src/api.js (hunk 2 above)
Comment (ryansloan) [reply src/api.js]: The blocks editor enforces uniqueness, I'm not sure how much of that is just the editor vs. some other underlying requirement

**Pull request merged**